Aymanprojects/basic-authentication 1st Challenge Completed

AymanProjects opened the pull request:
### My attempt of completing the challenge using best practices 🐦🦜

AymanProjects committed:
Clean Up The Template

AymanProjects committed:
Completed challenge #1

AymanProjects committed:
Fixing imports due to folders movements

AymanProjects committed:
Enhanced working with token

mutairibassam commented:
@AymanProjects you still need to develop some UIs. Please push the changes once it's done to this pull request. I will reopen the pull request.

AymanProjects committed:
Added UI

AymanProjects commented:
UI has been added. 
You can review the code now :)

AymanProjects committed:
Changes:
- Added Snack Bars
- Keys moved to the constants folder

AymanProjects committed:
Fixes:
- Fixed an issue where user could login without authenticaton using old local data

mutairibassam commented:
@AymanProjects please pull the latest changes that's been added in main branch and create another pull request. New changes will add github action for flutter.

AymanProjects committed:
Merge branch 'bepitome:main' into aymanprojects/basic-authentication

AymanProjects commented:
@mutairibassam  Changes merged and committed here.

mutairibassam commented:
@AymanProjects please check the failure and update accordingly. You might need to check project test folder as it's failing with command `flutter test`.

AymanProjects committed:
Added Test Folder

AymanProjects commented:
Added Test Folder As Required

AymanProjects commented:
CI Failed ? You gotta be kidding me

mutairibassam commented:
@AymanProjects test failed again since there is no any test to be run. In addition you should run `flutter test` locally before pushing to ensure passing all the use-cases.

AymanProjects committed:
Inluded Test case inside test folder

AymanProjects commented:
Should work now

AymanProjects committed:
Changes:
* Added helper class around storage service

mutairibassam commented on lib/src/data/helper_classes/http_client.dart:
```diff
@@ -0,0 +1,67 @@
+    }
+  }
+
+  @override
+  Future<dynamic> post({
+    required String endPoint,
+    Map<String, dynamic>? query,
+  }) async {
+    final auth = locator<TokensService>();
+    Uri url = Uri.parse('$kApiBaseUrl$endPoint').replace(
+      queryParameters: query,
+    );
+    final headers = {
+      'Content-type': 'application/json',
+      'Accept': 'application/json',
+      'Authorization': await auth.getLocalAccessToken(),
+    };
+    final response = await http.post(url, headers: headers);
+    final body = jsonDecode(response.body);
+    final result = body['result'];
+    if (response.statusCode == 200) {
+      return result;
+    } else {
+      throw HttpException(
+        statusCode: response.statusCode,
+        message: result as String,
+      );
+    }
+  }
+}
```
In both `get` & `post` functions your response handling should cover more edge cases. 

Your success validation has only one condition if the statusCode is 200. As you know we have different statusCode for success and for failures for example 201 for successfully created and 202 for acceptance. For more details check [mozilla website](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status).

I've experimented one-case to prove the concept. I changed the statusCode to 201 from client side to simulate statusCode incompatibility since I can't change it from server side.

As you can see it throws an exception even though I received a valid token.

![Screen Shot 1444-02-07 at 1 18 45 PM](https://user-images.githubusercontent.com/54819958/188266610-13b438c6-bc6d-4ffa-8d17-73376e9d38a0.png)

And the notifier notify the user with a technical error which need to be handled as well.

![Screen Shot 1444-02-07 at 1 20 42 PM](https://user-images.githubusercontent.com/54819958/188266686-aedacbba-0e0b-480e-8165-2e9de6d191a4.png)

AymanProjects replied:
This should be done in coordination between the back end & front end developers. 
You should list all the status code that you use in your backend.

AymanProjects replied:
For the notification, you are write, I'm just throwing everything at the user 😅

mutairibassam replied:
I put my comments to tell you about specific things related to technical and soft skills. The good thing is you directly come to the point.

The takeaway here is about Backend and Frontend communication. You should enforce/convince backend developers to change backend code to minimize frontend logic and avoid hardcoding values. Believe you have to `push` them so hard. 

I know you already did #3 but didn't push me to change. It might be only a challenge and no harm to release with this code but never do it in real life.

mutairibassam commented on lib/src/presentation/views/sign_in/sign_in_viewmodel.dart:
```diff
@@ -0,0 +1,57 @@
+    return formKey.currentState?.validate() ?? false;
+  }
+
+  String? usernameValidator(String? value) {
+    if (value is! String || value.isEmpty) {
+      return 'Username can\'t be empty';
+    }
+    return null;
+  }
+
+  String? passwordValidator(String? value) {
+    if (value is! String || value.isEmpty) {
+      return 'Password can\'t be empty';
+    }
+    return null;
+  }
```
I believe the condition might be optimized to avoid one unneeded check. The method always expects `String` and will complain if you pass `non-string` type.

In this validation we have to understand two aspects, form value and validator value. Since you are using `TextEditingController()` the value will be what the user have typed or null. In null case, we can see there is a logic to override the value to be empty string.

![Screen Shot 1444-02-07 at 2 20 14 PM](https://user-images.githubusercontent.com/54819958/188268056-f42f49ee-9737-4c6a-a7a2-f683550d40b6.png)

![Screen Shot 1444-02-07 at 2 15 21 PM](https://user-images.githubusercontent.com/54819958/188268095-dfa9d21a-e7bd-4fd2-b00d-da2ca10058d6.png)


---
So, we don't need to validate the type anymore and to make sure you might check the type at runtime and it will always be `String`.

```dart
  String? usernameValidator(String? value) {
    /// run time
    var runtimeType = value.runtimeType;
    
    if (value is! String || value.isEmpty) {
      return 'Username can\'t be empty';
    }
    return null;
  }
```

An optimized condition might be similar to the below:
```suggestion
  String? usernameValidator(String? value) {
    if (value!.isEmpty) {
      return 'Username can\'t be empty';
    }
    return null;
  }

  String? passwordValidator(String? value) {
    if (value!.isEmpty) {
      return 'Password can\'t be empty';
    }
    return null;
  }
```

Please let me know if you've considered edge cases that didn't know about.

AymanProjects replied:
I think your code will break if Flutter Team decided to pass a nullable value in the future. 
I always try to avoid `!`

mutairibassam replied:
As you know I am not Flutter guy. I put `!` just because the complier yelled at me. I would love to know why are you trying to avoid it?

AymanProjects replied:
Dart is a null-safe language. The complier will always tell you if a value might be null. 

if a variable has a type of `String` that means it can never be null, therefore you don't need to write an `if` statement to check if it is null or not.  

if a variable has a type of `String?`   that means it could be a `String` or `null`, therefore the compiler will force you to write an `if` statement to check if it is null or not. 

There are cases when you define a variable as `String?` but at some point, you know that it will not be null, that's where you can use `!`, to tell the compiler that I'm 100% sure that this value won't be null.

AymanProjects replied:
It a good practice to avoid `!` and do what the compiler tells you.

mutairibassam commented on lib/src/services/auth_service.dart:
```diff
@@ -0,0 +1,47 @@
+    return token.isNotEmpty;
+  }
+
+  Future<void> login(String username, String password) async {
+    // Do basic auth
+    String base64Encoded = base64.encode(utf8.encode('$username:$password'));
+    await tokens.updateAccessTokenLocally('Basic $base64Encoded');
```
It seems you are storing the encoded value in local storage then you override it with the valid token if received.

I would recommend to separate the values to have each one its own local key.

AymanProjects replied:
Separation has no advantage here, because you will never need both tokens at the same time.

mutairibassam replied:
For code scalability, and readability you should separate them.

mutairibassam commented on lib/src/presentation/views/home/home_view.dart:
```diff
@@ -0,0 +1,43 @@
+          child: Row(
+            mainAxisAlignment: MainAxisAlignment.center,
+            children: [
+              Column(
+                mainAxisAlignment: MainAxisAlignment.center,
+                children: const [
+                  HomeTitle(),
+                  LargeSpace(),
+                  HomeUserDetails(),
+                  LargeSpace(),
+                  HomeTeammatesButton(),
+                  SmallSpace(),
+                  HomeLogoutButton(),
+                ],
+              ),
```
Your home widget design will cause RenderFlex overflow exception for some mobile screen sizes. You can validate your design using Widget Inspector. I faced it myself using Samsung S9.

![Screen Shot 1444-02-07 at 2 53 23 PM](https://user-images.githubusercontent.com/54819958/188269299-e59b180b-b46a-4b46-a0c1-788c09a42567.png)

Also, consider to go through the below references:
- https://stackoverflow.com/questions/49480051/flutter-dart-exceptions-caused-by-rendering-a-renderflex-overflowed
- https://github.com/flutter/flutter/blob/e3005e6962cfefbc12e7aac56576597177cc966f/packages/flutter/lib/src/widgets/basic.dart#L3724
- https://docs.flutter.dev/codelabs/layout-basics

mutairibassam reviewed (changes requested):
First review iteration.

AymanProjects committed:
Organizing code

AymanProjects committed:
Fixed typos

AymanProjects committed:
Readable exception messages

AymanProjects committed:
Code refactor

AymanProjects committed:
Refactoring local storage

AymanProjects committed:
Removed unwanted code

AymanProjects committed:
re arranged file locations

AymanProjects committed:
Re-archetcture
